lxd/device/nic_ipvlan.go

```diff
@@ -0,0 +1,208 @@
+package device
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/lxc/lxd/lxd/instance"
+	"github.com/lxc/lxd/shared"
+)
+
+type nicIPVLAN struct {
+	deviceCommon
+}
+
+func (d *nicIPVLAN) CanHotPlug() (bool, []string) {
+	return false, []string{}
+}
+
+// valite checks the supplied config for correctness.
+func (d *nicIPVLAN) validate() error {
+	if d.instance.Type() != instance.TypeContainer {
+		return ErrUnsupportedDevType
+	}
+
+	extensions := d.state.OS.LXCFeatures
+	if !extensions["network_ipvlan"] || !extensions["network_l2proxy"] || !extensions["network_gateway_device_route"] {
+		return fmt.Errorf("Requires liblxc has following API extensions: network_ipvlan, network_l2proxy, network_gateway_device_route")
```

Comment:
Those kind of checks are better done on start than during validate.
I've removed a few recently as that effectively breaks folks using another system as their backup machine because of lack of kernel support for some feature or liblxc support for it.

We shouldn't allow known broken configurations (bad IP and such) but if we're just missing something on the machine at the time, we should allow it and error out on start.

Reply:
I suspect the issue here is that validate() is called on instantiation and so you don't yet know if you're going to start the container or what action will be performed. We don't care about host configuration for stop or delete, but we do for Start/Update.

So I suspect you'll want to split those checks into a pure logical config validation which can be done every time we instantiate a device and an environment validation which is only performed on start/update.

Reply:
the macvlan case looks much closer to what I'd expect here

Reply:
@stgraber agreed, I've renamed `validate()` to `validateConfig()` and then added a new `validateEnvironment()` function to all device types that is called from inside `Start()` and `Update()` (if running).